### PR TITLE
Expose the current FRR status

### PR DESCRIPTION
**Pull request by fedepaol**
We add a new CRD which will expose the following information:

- the current running configuration of an frr instance
- the desired configuration that the daemon wants to apply
- the result of the last desired configuration apply, which might be
  success or the failure message

### CODE REVIEW

File: internal/frr/frr.go [hunk 1]
```diff
@@ -45,67 +62,110 @@ func (f *FRR) ApplyConfig(config *Config) error {
...
+			case <-ctx.Done():
+				return
+			}
 		}
 	}()
 }
 
-const statusFileName = "/etc/frr_reloader/.status"
+const (
+	statusFileName    = "/etc/frr_reloader/.status"
```
Comment (oribon) [inline internal/frr/frr.go:119]: it's a bit confusing now that this file is called status when what it does is holding a timestamp, can you change it?

File: internal/controller/frrstatus_controller.go
Comment (oribon) [inline internal/controller/frrstatus_controller.go:1]: nit: rename to `frrnodestate_controller.go`

**Review by oribon — COMMENTED** [review]
went over ~2 commits, will continue later

File: internal/frr/frr.go [hunk 2]
```diff
@@ -45,67 +62,110 @@ func (f *FRR) ApplyConfig(config *Config) error {
...
-	if len(lastReloadStatus) != 2 {
-		level.Error(l).Log("op", "reload-validate", "cause", "Fields", "bytes", string(bytes))
-		return
+	res := Status{
+		updateTime: timeStamp,
+	}
+
+	bytes, err := os.ReadFile(lastAppliedResult)
+	if err != nil && !os.IsNotExist(err) {
+		return Status{}, fmt.Errorf("failed to read running config file: %w", err)
```
Comment (oribon) [inline internal/frr/frr.go:137]: nit: failed to read last error file

File: internal/frr/frr.go [hunk 3]
```diff
@@ -45,67 +62,110 @@ func (f *FRR) ApplyConfig(config *Config) error {
...
+	bytes, err = os.ReadFile(runningConfig)
+	if err != nil && !os.IsNotExist(err) {
+		return Status{}, fmt.Errorf("failed to read running config file: %w", err)
 	}
+	res.Current = string(bytes)
 
-	*prevReloadTimeStamp = timeStamp
+	bytes, err = os.ReadFile(lastAppliedConfig)
+	if err != nil && !os.IsNotExist(err) {
+		return Status{}, fmt.Errorf("failed to read last appl file: %w", err)
```
Comment (oribon) [inline internal/frr/frr.go:152]: nit: last applied

File: internal/frr/frr.go [hunk 4]
```diff
@@ -45,67 +62,110 @@ func (f *FRR) ApplyConfig(config *Config) error {
...
+				status, err := fetchStatus()
+				if err != nil {
+					level.Error(l).Log("op", "fetch status", "error", err)
+					break
+				}
+				if status.updateTime == f.Status.updateTime {
+					break
+				}
+				if status.LastReloadResult != ReloadSuccess {
+					level.Error(l).Log("op", "fetch status", "error", err)
```
Comment (oribon) [inline internal/frr/frr.go:102]: nit: `err` here is irrelevant

File: internal/frr/frr.go [hunk 5]
```diff
@@ -45,67 +62,110 @@ func (f *FRR) ApplyConfig(config *Config) error {
...
-			validateReload(l, &prevReloadTimeStamp, reload)
-		case <-ctx.Done():
-			return
+		for {
+			select {
+			case <-ticker.C:
+				status, err := fetchStatus()
+				if err != nil {
+					level.Error(l).Log("op", "fetch status", "error", err)
+					break
```
Comment (oribon) [inline internal/frr/frr.go:96]: could this suggest an error actually happened on frr's side? I think not but wonder if we should update the status here as well (or comment why we aren't)

File: internal/controller/frrstatus_controller_test.go [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+				return nodeStatusList.Items[0]
+			}, time.Minute, time.Second).Should(
+				gstruct.MatchFields(gstruct.IgnoreExtras, gstruct.Fields{
+					"ObjectMeta": gstruct.MatchFields(gstruct.IgnoreExtras, gstruct.Fields{
+						"Name":      Equal(testNodeName),
+						"Namespace": Equal(testNamespace),
+					}),
+					"Status": gstruct.MatchFields(gstruct.IgnoreExtras, gstruct.Fields{
+						"RunningConfig": Equal("foo"),
+						"DesiredConfig": Equal("bar"),
```
Comment (oribon) [inline internal/controller/frrstatus_controller_test.go:83]: missing lastReloadResult?

File: e2etests/tests/status.go [hunk 7]
```diff
@@ -0,0 +1,290 @@
...
+			framework.ExpectNoError(err)
+			ginkgo.By("Creating a configuration with no neighbors")
+			err = updater.Update([]corev1.Secret{}, frrconfig)
+			framework.ExpectNoError(err)
+
+			for _, node := range nodes {
+				Eventually(func() error {
+					return nodeMatchesStatus(cl, node.Name, func(status frrk8sv1beta1.FRRNodeState) error {
+						if status.Status.LastReloadResult != reloadSuccess {
+							return fmt.Errorf("LastReloadResult is not success")
```
Comment (oribon) [inline e2etests/tests/status.go:107]: formatting with the node's name might be helpful

File: internal/controller/frrstatus_controller.go [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+// +kubebuilder:rbac:groups="",resources=nodes,verbs=get;list;watch
+
+func (r *FRRStatusReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	level.Info(r.Logger).Log("controller", "FRRStatusReconciler", "start reconcile", req.NamespacedName.String())
+	defer level.Info(r.Logger).Log("controller", "FRRStatusReconciler", "end reconcile", req.NamespacedName.String())
+
+	frrStatus := r.FRRStatus.GetStatus()
+	if frrStatus == r.currentStatus { // Do nothing
+		return ctrl.Result{}, nil
+	}
```
[the remark this answers is not in the document]
Comment (fedepaol) [reply internal/controller/frrstatus_controller.go:84]: I will compare the actual status, will cost a get more but it's cleaner

File: api/v1beta1/frr_node_status.go [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// FRRNodeStateSpec defines the desired state of FRRNodeState.
+type FRRNodeStateSpec struct {
+}
+
+// FRRNodeStateStatus defines the observed state of FRRNodeState.
+type FRRNodeStateStatus struct {
```
[the remark this answers is not in the document]
Comment (fedepaol) [reply api/v1beta1/frr_node_status.go:28]: yep I will do in a different pr

File: internal/frr/frr.go (hunk 5 above)
Comment (fedepaol) [reply internal/frr/frr.go:96]: this is more an infrastructure failure, where we were not able to fetch the information and thus we are not able to understand if the status is failed or not. Let me add a comment.

**Review by oribon — COMMENTED** [review]
few nits, other than that lgtm

File: internal/controller/frrstate_controller.go [hunk 10]
```diff
@@ -0,0 +1,139 @@
...
+func (r *FRRStateReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	level.Info(r.Logger).Log("controller", "FRRStateReconciler", "start reconcile", req.NamespacedName.String())
+	defer level.Info(r.Logger).Log("controller", "FRRStateReconciler", "end reconcile", req.NamespacedName.String())
+
+	status := &frrk8sv1beta1.FRRNodeState{}
+
+	err := r.Client.Get(ctx, types.NamespacedName{Name: r.NodeName, Namespace: r.Namespace}, status)
+	if k8serrors.IsNotFound(err) {
+		status.Name = r.NodeName
+		status.Namespace = r.Namespace
```
Comment (oribon) [inline internal/controller/frrstate_controller.go:88]: nit: uneeded now

**Comment by fedepaol** [comment]
fixed the nits, sending to merge

**Pull request merged**